### ISSUE TITLE
Shift + Click in AE crafting terminal causes severe lag / server kick for some or all players

### ISSUE THREAD

**Issue by Devtron3D** [GTNewHorizons/GT-New-Horizons-Modpack#6570]
#### Which modpack version are you using?
2.0.9.0
#
#### If in multiplayer; On which server does this happen? 
Prospercraft
#
#### What did you try to do?
Crafting tiny dust to normal dust in the AE crafting terminal with shift + mouse left click

#
#### What happened ? (Attach screenshots if needed)
![image](https://user-images.githubusercontent.com/3405825/92011465-615c7200-ed4b-11ea-844f-3017500c081c.png)

When Shift + mouse left click,  it causes SEVERE server lag, to the point of kicking people off the server, and TPS freeses for a long time.

I know  it was a bug in the wireless crafting terminaland, in the last update from start of the year (i reported that). But i was unable to replicate it in the standart crafting terminal (one connected to an AE fluix cable) , after this 2.0.9.0 update i was just now crafting some indium to dust by hand and people got kicked, retested it and it causes people to get kicked and 10-20+ seconds lag/tps stop

#
#### What do you suggest instead/what changes do you propose?
If this cannot be fixed, maybe remove the ability to shift click in AE terminals

**Comment by repo-alt** [comment]
Can't reproduce that on Eta and locally

**Comment by Devtron3D** [comment]
Ok, dont know if it is item count in ME system, maybe drive defragmentation,  did test on ETA, doing it when i had 15k wood and making planks, droped the tps (bit hard to tell with that many on)  and caused a tiny frese . but on prospercraft doing it with 184k tiny cinibar causes major lag
![image](https://user-images.githubusercontent.com/3405825/92019873-95d62b00-ed57-11ea-8c2e-7e8069a0e018.png)

going to see if i can get a copy of my void to do single player tests

**Comment by Devtron3D** [comment]
![image](https://user-images.githubusercontent.com/3405825/92019983-bbfbcb00-ed57-11ea-960f-14007e305eb0.png)
ME current content sorted by number of items

**Comment by boubou19** [comment]
I can produce major lag on delta by doing it in my gigantic AE storage (storage buses on 3 infinity chests + ~40 cells). So yeah it may be linked to the amount of stored items.

**Comment by repo-alt** [comment]
In this case just extracting large amounts of items from AE would cause lag

**Comment by Cleista** [comment]
Massive lag on my server crafting tiny dusts as well, about 140 drives and several filtered storage buses on drawer controllers and one unfiltered storage bus on a ore miner subnetwork. 

Problem was somewhat solved by using amazon packager but still recommend seeing if it's possible to optimize at all, the oredict export bus doesn't lag so why would manually crafting dust freeze server?

**Comment by Devtron3D** [comment]
ok i just tested my garden dimention on a single player world, it causes several seconds of lag, so it is tied to the size of the me system and or the amount of items in it

**Comment by repo-alt** [comment]
Can you compare Shift click craft lag in the Wireless crafting terminal and usual crafting terminal?

**Comment by Cleista** [comment]
Wireless terminal feels slightly worse (about 2x longer response time), but both of them tank TPS.

**Comment by NoX-programer** [comment]
@repo-alt I'm sure, to reproduce it, you need to have really big quantities of source materials going to craft (here: dusts) and maybe also large AE network with maybe lot of ME drives stuffed. It is long ago known problem of AE growing from fact AE cannot calculate whole stack at once because there is some special recipes (using GT tools that loses durability, for example) therefore it craft 64 times when shift+click causing described lags. If players want not to lag server it's only the workaround option to craft slowly and better automatically.

**Comment by Devtron3D** [comment]
It might be a workaround to ask people to stop shift + clicking a stack
But remember many wont, and will ignore it just to craft faster.
Either the shift + click needs to be disabled or a solution for why when crafting something as simple as tiny dust, or wood planks, causes so much lag needs to be found
I also wonder how much lag it introduces just single clicking fast to craft the same does.

**Comment by repo-alt** [comment]
Technically it is possible to check if there are items with durability in the recipe and if not then try to craft whole stack

**Comment by repo-alt** [comment]
(dubarility and NBT)

**Comment by NoX-programer** [comment]
This will solve the problem, I'm sure (still it will be possible to knowingly lag the server with complicated recipes, but it will be definitely ban). Will you fix AE2 code or what? BTW, at the same time, it will be helpful to catch that case when someone is lagging the server by shift+crafting and log it to logs&server_console.

**Comment by repo-alt** [comment]
Crafting recipe may have special handlers? What do you mean?

**Comment by NoX-programer** [comment]
Rare case. Example from my script.zs:
```
var dra_arch = [<draconicevolution:dislocator_advanced>, <draconicevolution:draconic_helm>, <draconicevolution:draconic_chest>, <draconicevolution:draconic_legs>, <draconicevolution:draconic_boots>, <draconicevolution:draconic_axe>, <draconicevolution:draconic_bow>, <draconicevolution:draconic_hoe>, <draconicevolution:draconic_pick>, <draconicevolution:draconic_shovel>, <draconicevolution:draconic_staff_of_power>, <draconicevolution:draconic_sword>] as IItemStack[];
for i, curitem in dra_arch {
    recipes.addShapeless(curitem, [curitem.marked("same")], function(output, inputs, crafting) { return inputs.same; });
}
```

**Comment by repo-alt** [comment]
That would just translate in a few wildcard shapeless recipes. Anyway, either IRecipe or crafting pattern does not have any special information.

**Comment by NoX-programer** [comment]
Recipe with not just input and output items, but with function existing, which can actually doing something "special". Better to avoid crafting whole stacks in such case to prevent dupes, glitches, etc.

**Comment by Glease** [comment]
I'm still not convinced this is somehow related to crafting being slow. If you have problems here, then you should equally have problem in a normal crafting bench - which you don't. I still think it's because the AE2's slow IO speed. I believe it is doing 9*64 (9 extraction per craft, 64 craft per click) transcations per shift click (which is 9*64*4=2304 QPS under 240 APM). Your usual export bus do less than 1QPS even with 4 acceleration card because it operates fully in bulk mode, that is, one stack per operation instead of one item per operation.

I'd say this is really too much for even the most optimized IGrid implementation.

**Comment by Glease** [comment]
FYI The related code that actually invoke crafting handler and move the items are located in this file: https://github.com/GTNewHorizons/Applied-Energistics-2-Unofficial/blob/rv3-1.7.10/src/main/java/appeng/container/slot/SlotCraftingTerm.java

**Comment by Glease** [comment]
My opinion is, let's do one or two craft per tick, and let the 64 craft happens throughout the coming 3.2/1.6 second. This does make it slower but it reduce the peak server load to ~1.5%/3% of original.

**Comment by repo-alt** [comment]
Grid extraction does not really care how much do you extract, it is updating the grid after the extraction is costly, so extracting nine stacks should be much faster than extracting 64*9 items

**Comment by repo-alt** [comment]
Refactoring the crafting slot is of course very unappealing task but it seems to be doable

**Comment by Glease** [comment]
The problem is you have to change how crafting fundamentally happens, so it doesn't do 9\*64 extractions. You could try to fill the crafting grid beforehand, and push the extra stuff back afterwards, but there are stuff like water buckets out there. Right now I don't have a solid plan on how this refactoring should happen... Maybe we could bend the behavior a bit to get better performance?

**Comment by repo-alt** [comment]
Well, fist of all - do not apply optimization if items with NBT are in the pattern. Then try to fill the crafting grid with the whole stacks  of the pattern inputs. If that worked - ok, do batch crafting until it breaks, then fallback

**Comment by NoX-programer** [comment]
I thinking and cannot understand - why items with NBT are bad, how them guilty? If them prove not by good stacking and filling our crafting grid so why not? Maybe it's the situation the player have a lot of them, so let's craft them!

**Comment by repo-alt** [comment]
If they do stack then is is probably ok, but it is usually not the case.

**Comment by NoX-programer** [comment]
So let the stacking alhorithm do it's job, and we will just rely on it.

**Comment by KaneHart** [comment]
This the same issue with Crafting Tables, etc I reported ages ago btw. But I think I remember debugging this years ago with the older version like 1.7.10 caused a Network thread stall and right up the butt you get it.

idk if they solved this later on if I have energy later on I will see if I can find some code for it.

**Comment by repo-alt** [comment]
Seriously

**Comment by KaneHart** [comment]
> Seriously

So I looked it up found the right time and such and it was "apparently" Patched near the end of 1.7.10. The other major crafting issue I saw was storage drawers and controller using lots of storage drawers. Not sure if your using this method or not. I will try profile it once I get there myself.

**Comment by GTNH-Afx237v7** [comment]
This has now lead to not being able to craft more than a single stack of output materials, when one of the slots has more than one item in it.
Example: I have 56,000 tiny piles in a small AE network. Filling the crafting grid and shift-left clicking with 8 of the slots containing one tiny pile and the 9th slot containing 2 tiny piles will craft 64 full dusts and leave 8 tiny piles in the crafting grid, so you can’t shift-click again.
As long as each slot has only one item you can shift-click just fine and keep getting full stacks as output.

**Comment by repo-alt** [comment]
Don't do that then? As in - can you provide more valid use case?

**Comment by GTNH-Afx237v7** [comment]
Assuming you pick up a full stack and then drag it across the GUI so that it will be evenly distributed across all slots - I assume people do that? At least I do this when batch crafting.

There is however a duplication issue now:
I started with 244 stone dust in a small test setup. Turning them into tiny piles works fine. If you then populate the grid as mentioned above (i.e. 8 slots with one tiny pile and the 9th slot with 2 tiny piles) , then shift click it will craft 64 full dusts and leave 8 tiny piles in the GUI. Drop those 64 dusts back into the system, repopulate the empty slot with 2 tiny piles and repeated. Eventually I ended up with 271 full dusts and 5 tiny piles.
Also tested with shapeless crafting of HSS-G dust (using 60 tungstensteel, 12 vanadium, 12 chrome and 24 molybdenum dust):
Each slot will get one dust, except for chrome which will be populated with 2 dusts. Shift click, repopulate the empty slot for chrome dust and end up with 96 HSS-G dust and 6 extra chrome (where 96 HSS-G and no extra dusts were expected).
